app/controllers/api/v1/migrations/external_controller.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    module Migrations
+      class ExternalController < ApiController
+        skip_before_action :verify_authenticity_token
+        skip_before_action :ensure_authenticated
+
+        rescue_from ActiveSupport::MessageEncryptor::InvalidMessage do |exception|
+          log_exception exception
+          render_error errors: [Rails.configuration.custom_error_msgs[:missing_params]], status: :bad_request
+        end
+
+        # POST /api/v1/migrations/roles.json
+        # Expects: { role: { :name } }
+        # Returns: { data: Array[serializable objects(recordings)] , errors: Array[String] }
+        # Does: Creates a role.
+        def create_role
+          role = Role.new role_params.merge(provider: 'greenlight')
+
+          return render_error status: :bad_request unless role.save
+
+          render_data status: :created
+        end
+
+        private
+
+        def role_params
+          decrypted_params.require(:role).permit(:name)
+        end
+
+        def decrypted_params
+          encrypted_params = params.require(:v2).require(:encrypted_params)
+
+          raise ActiveSupport::MessageEncryptor::InvalidMessage unless encrypted_params.is_a? String
+
+          crypt = ActiveSupport::MessageEncryptor.new(Rails.application.secrets.secret_key_base[0..31])
+          decrypted_params = crypt.decrypt_and_verify(encrypted_params) || {}
+
+          raise ActiveSupport::MessageEncryptor::InvalidMessage unless decrypted_params.is_a? Hash
+
+          ActionController::Parameters.new(decrypted_params)
+        end
```

Comment:
The `decrypted_params` method is an API for validating and decrypting the encrypted request parameters while it can be used in place of the regular `params` method.
And it:
1. Validates the structure of the request params, any valid request should respect the schema `:v2 => {:encrypted_params => String}`
2. Extracts the `encrypted_params` string if 1. passes.
3. Decrypts and verifies the `encrypted_params` ciphertext.
4. If 3. passes it will further verify the serialized plaintext `decrypted_params` is a `Hash`.
5. It will form in an `ActionController::Parameters` instance out of the `decrypted_params` hash.

Output:
1. It can raise `ActiveSupport::MessageEncryptor::InvalidMessage` and `ActionController::ParameterMissing` (in both cases the v3 will log the exception  and render_error a :bad_request) for invalid inputs.
2. It returns a params instance of the decrypted params or an empty hash `{}`.

Reply:
So one can add in a new action and just use the `decrypted_params` in place of the regular `params` method which will enable the action with cryptographic verification and decryption while providing a regular params object seamlessly.